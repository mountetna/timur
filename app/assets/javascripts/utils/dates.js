const pad = (v) => ('00' + v).slice(-2);

export const formatDate = (date)=>{
  if(!date) return null;
  if(!(date instanceof Date)) date = new Date(date);

  let day = pad(date.getDate());
  let month = pad(date.getMonth());
  let year = pad(date.getFullYear());

  return `${month}-${day}-${year}`;
};

export const formatTime = (date)=>{
  if(!date) return null;
  if(!(date instanceof Date)) date = new Date(date);
  let hours = pad(date.getHours());
  let minutes = pad(date.getMinutes());
  return `${hours}:${minutes}`;
};
