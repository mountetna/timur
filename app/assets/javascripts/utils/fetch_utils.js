import downloadjs from 'downloadjs';

export const checkStatus = (response)=>{
  if(response.status >= 200 && response.status < 300){
    return response;
  }
  else{
    let error = new Error(response.statusText);
    error.response = response;
    throw error;
  }
};

export const parseJSON = (response)=>{
  try{
    return response.json();
  }
  catch(err){
    console.log(err);
  }
};

export const makeBlob = (response)=>{
  return response.blob();
};

export const generateDownload = (filename)=>{
  return (blob)=>{
    return downloadjs(blob, filename, blob.type);
  }
};

export const headers = (...types)=>{
  let _headers = {};
  let add = (header, value)=> _headers[header] = value;

  for(let type of types){
    switch(type){
      case 'json':
        add( 'Content-Type', 'application/json');
        break;
      case 'csrf':
        let csrf = document.querySelector('meta[name=csrf-token]');
        if(csrf) add('X-CSRF-Token', csrf.getAttribute('content'));
        break;
      default:
        break;
    }
  }

  return _headers;
};
