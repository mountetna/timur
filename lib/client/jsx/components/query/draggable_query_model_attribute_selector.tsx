import React from 'react';

import {Draggable} from 'react-beautiful-dnd';

import QueryModelAttributeSelector from './query_model_attribute_selector';

const DraggableQueryModelAttributeSelector = (props: any) => {
  return (
    <Draggable
      draggableId={`column-${props.columnIndex}`}
      index={props.columnIndex}
    >
      {(provided: any) => (
        <div
          ref={provided.innerRef}
          {...provided.draggableProps}
          {...provided.dragHandleProps}
        >
          <QueryModelAttributeSelector {...props} />
        </div>
      )}
    </Draggable>
  );
};

export default DraggableQueryModelAttributeSelector;
