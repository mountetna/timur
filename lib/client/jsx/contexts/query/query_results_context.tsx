import React, {useState, createContext, useCallback} from 'react';

import {QueryResponse} from './query_types';

export const defaultQueryResultsParams = {
  expandMatrices: true,
  flattenQuery: true,
  page: 0,
  pageSize: 10,
  data: {} as QueryResponse,
  numRecords: 0,
  queryString: ''
};

const defaultQueryResultsState = {
  ...defaultQueryResultsParams
};

export type QueryResultsState = Readonly<typeof defaultQueryResultsState>;

export const defaultQueryResultsContext = {
  state: defaultQueryResultsState as QueryResultsState,
  setExpandMatrices: (expandMatrices: boolean) => {},
  setFlattenQuery: (flattenQuery: boolean) => {},
  setPage: (page: number) => {},
  setPageSize: (pageSize: number) => {},
  setDataAndNumRecords: (data: QueryResponse, numRecords: number) => {},
  setQueryString: (queryString: string) => {},
  setResultsState: (newState: QueryResultsState) => {}
};

export type QueryResultsContextData = typeof defaultQueryResultsContext;

export const QueryResultsContext = createContext(defaultQueryResultsContext);
export type QueryResultsContext = typeof QueryResultsContext;
export type ProviderProps = {params?: {}; children: any};

export const QueryResultsProvider = (
  props: ProviderProps & Partial<QueryResultsContextData>
) => {
  const [state, setState] = useState(
    props.state || defaultQueryResultsContext.state
  );

  const setExpandMatrices = useCallback(
    (expandMatrices: boolean) => {
      return {
        ...state,
        expandMatrices
      };
    },
    [state]
  );

  const setFlattenQuery = useCallback(
    (flattenQuery: boolean) => {
      setState({
        ...state,
        flattenQuery
      });
    },
    [state]
  );

  const setPage = useCallback(
    (page: number) => {
      setState({
        ...state,
        page
      });
    },
    [state]
  );

  const setPageSize = useCallback(
    (pageSize: number) => {
      setState({
        ...state,
        pageSize
      });
    },
    [state]
  );

  const setDataAndNumRecords = useCallback(
    (data: QueryResponse, numRecords: number) => {
      setState({
        ...state,
        data,
        numRecords
      });
    },
    [state]
  );

  const setQueryString = useCallback(
    (queryString: string) => {
      setState({
        ...state,
        queryString
      });
    },
    [state]
  );

  const setResultsState = useCallback((newState: QueryResultsState) => {
    setState({
      ...newState
    });
  }, []);

  return (
    <QueryResultsContext.Provider
      value={{
        state,
        setExpandMatrices,
        setFlattenQuery,
        setPage,
        setPageSize,
        setDataAndNumRecords,
        setQueryString,
        setResultsState
      }}
    >
      {props.children}
    </QueryResultsContext.Provider>
  );
};
