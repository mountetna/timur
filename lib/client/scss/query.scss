@import 'globals';

.query-map {
  display: flex;
  margin-top: 10px;
  position: relative;
  font-family: $body_font;

  .map {
    flex-basis: 600px;
    border-right: 1px solid lightgray;

    .heading {
      position: absolute;
      left: 15px;
      top: 20px;
      border-bottom: $thin_rule;
    }
    svg {
      overflow: visible;
      line {
        stroke-width: 3px;
        stroke: #aca;
      }
      #arrow {
        fill: #aca;
      }
    }
    .model_node {
      cursor: pointer;
      position: absolute;

      font-family: $body_font;
      font-size: 0.75em;

      border-radius: 3px;
      border: 2px solid #aaa;
      background: #f2f2f2;
      box-shadow: 0 0 2px 0 #ccc;
      padding: 5px;

      &.selected {
        border: 2px solid #0c0;
        background: #cfc;
      }

      &.disabled {
        background: darkgray;
        border: 2px solid #aaa;
        cursor: not-allowed;

        &:hover {
          background: lightpink;
        }
      }

      transform: translate(-50%, -50%);
      &:hover {
        background: #cec;
      }
    }
  }
}

.query-attributes-modal {
  width: 800px;
  padding: 20px;
  border-radius: 2px;
  background: white;
  max-height: 70vh;
  font-family: $body_font;
  overflow-y: auto;

  .actions {
    text-align: right;

    button {
      @include green_button;
    }
  }
}
