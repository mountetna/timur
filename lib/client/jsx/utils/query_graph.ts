// Construct a query graph from the Magma Models,
//   so we can traverse and ask it questions.
import {DirectedGraph} from 'etna-js/utils/directed_graph';
import {Model, Template, Attribute} from '../models/model_types';

export class QueryGraph {
  models: {[key: string]: Model};
  graph: DirectedGraph;
  unallowedModels: string[] = ['project'];
  includedLinkTypes: string[] = ['link', 'table'];
  allowedModels: Set<string>;
  initialized: boolean = false;

  constructor(magmaModels: {[key: string]: Model}) {
    this.models = magmaModels;
    this.graph = new DirectedGraph();
    this.allowedModels = new Set();

    // We ignore the project model and any links
    //   to project, for querying purposes only.
    // Also include linked models, to capture those paths.
    Object.entries(magmaModels).forEach(
      ([modelName, modelDefinition]: [string, Model]) => {
        if (this.unallowedModels.includes(modelName)) return;

        let template: Template = modelDefinition.template;
        this.allowedModels.add(modelName);

        if (!this.unallowedModels.includes(template.parent))
          this.graph.addConnection(template.parent, modelName);

        Object.values(template.attributes)
          .filter(
            (attr: Attribute) =>
              this.includedLinkTypes.includes(attr.attribute_type) &&
              attr.link_model_name
          )
          .forEach((link: Attribute) => {
            if (link.link_model_name) {
              this.allowedModels.add(link.link_model_name);
              this.graph.addConnection(modelName, link.link_model_name);
            }
          });
      }
    );

    this.initialized = true;
  }

  template(modelName: string): any {
    if (!Object.keys(this.models).includes(modelName)) return null;

    return this.models[modelName].template;
  }

  pathsFrom(modelName: string): string[][] {
    // this.allowedModels could have disconnected models that
    //   where only connected to models in the unallowedModels list,
    //   so they won't appear in the graph, but the user may query on
    //   them and we have to account for that.
    // An immediate example is "document".
    if (!Object.keys(this.graph.children).includes(modelName)) return [];
    return this.graph.pathsFrom(modelName);
  }

  asNormalizedHash(modelName: string): {[key: string]: string[]} {
    // this.allowedModels could have disconnected models that
    //   where only connected to models in the unallowedModels list,
    //   so they won't appear in the graph, but the user may query on
    //   them and we have to account for that.
    // An immediate example is "document".
    if (!Object.keys(this.graph.children).includes(modelName)) return {};
    return this.graph.asNormalizedHash(modelName);
  }

  // Here we calculate parent paths as separate entities, instead
  //   of allowing them to be in a single, flattened array.
  parentPaths(modelName: string): string[][] {
    if (!Object.keys(this.graph.parents).includes(modelName)) return [];

    let results: string[][] = [];

    Object.keys(this.graph.parents[modelName]).forEach((p: string) => {
      if (p !== modelName) {
        let innerPaths = this.parentPaths(p);
        if (innerPaths.length === 0) {
          results.push([p]);
        } else {
          innerPaths.forEach((parentPath: string[]) => {
            parentPath.unshift(p);
            results.push(parentPath);
          });
        }
      }
    });

    return results;
  }

  allPaths(modelName: string | null): string[][] {
    if (!modelName) return [];

    if (!Object.keys(this.graph.children).includes(modelName)) return [];

    let parentPaths = this.parentPaths(modelName);

    // Any model that you can traverse to from any parent should
    //   also count as a path.

    // Children paths
    return this.pathsFrom(modelName)
      .concat(
        // paths up the tree
        parentPaths
      )
      .concat(
        // paths routing up through parents then down
        parentPaths
          .map((parentPath: string[]) =>
            parentPath
              .map((p: string, index: number) =>
                this.pathsFrom(p).map((path) =>
                  parentPath.slice(0, index).concat(path)
                )
              )
              .flat(1)
          )
          .flat(1)
      );
  }

  modelHasAttribute(modelName: string, attributeName: string) {
    if (!this.models[modelName]) return false;

    return !!this.models[modelName].template.attributes[attributeName];
  }

  stepIsOneToMany(start: string, end: string) {
    // For a single model relationship (start -> end),
    //   returns `true` if it is a one-to-many
    //   relationship.
    if (!this.modelHasAttribute(start, end)) return false;

    return ['table', 'collection'].includes(
      this.models[start].template.attributes[end].attribute_type
    );
  }

  attributeIsFile(modelName: string, attributeName: string) {
    if (!this.modelHasAttribute(modelName, attributeName)) return false;

    return ['file', 'image', 'file_collection'].includes(
      this.models[modelName].template.attributes[attributeName].attribute_type
    );
  }

  shortestPath(rootModel: string, targetModel: string): string[] | undefined {
    let paths = this.allPaths(rootModel).filter((potentialPath: string[]) =>
      potentialPath.includes(targetModel)
    );

    if (0 === paths.length) return;

    // Calculate steps to targetModel for each path
    let numberOfSteps = paths.map((path: string[]) => {
      return path.indexOf(targetModel);
    });
    let path = paths[numberOfSteps.indexOf(Math.min(...numberOfSteps))];

    // Direct children paths include the root, and
    //   we'll filter it out so all paths do not
    //   include the root model (eliminate redundancy).
    const pathWithoutRoot = path
      ?.slice(0, path.indexOf(targetModel) + 1)
      .filter((m) => m !== rootModel);

    return pathWithoutRoot;
  }

  sliceableModelNamesInPath(startModel: string, endModel: string): string[] {
    let modelsInPath = this.shortestPath(startModel, endModel);
    let previousModelName = startModel;
    let selectableModels: string[] = [];

    modelsInPath?.forEach((modelName) => {
      if (this.stepIsOneToMany(previousModelName, modelName)) {
        selectableModels.push(modelName);
      }
      previousModelName = modelName;
    });

    return selectableModels;
  }

  neighbors(modelName: string): {[key: string]: boolean} {
    // Returns the neighboring models (1 step away), with a boolean flag
    //   for one-to-many relationships. Includes original model.
    let results: {[key: string]: boolean} = {
      [modelName]: false
    };

    Object.keys(this.graph.children[modelName] || {}).forEach(
      (childNeighbor: string) => {
        results[childNeighbor] = this.stepIsOneToMany(modelName, childNeighbor);
      }
    );

    Object.keys(this.graph.parents[modelName] || {}).forEach(
      (parentNeighbor: string) => {
        results[parentNeighbor] = this.stepIsOneToMany(
          modelName,
          parentNeighbor
        );
      }
    );

    return results;
  }
}
