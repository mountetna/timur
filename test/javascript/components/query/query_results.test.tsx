import React from 'react';
import {render, fireEvent, waitFor, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import {mockStore, querySpecWrapper} from '../../helpers';
import QueryResults from '../../../../lib/client/jsx/components/query/query_results';
import {QueryGraph} from '../../../../lib/client/jsx/utils/query_graph';
import {defaultQueryResultsParams} from '../../../../lib/client/jsx/contexts/query/query_results_context';

const models = {
  monster: {
    documents: {},
    revisions: {},
    views: {},
    template: require('../../fixtures/template_monster.json')
  },
  prize: {
    documents: {},
    revisions: {},
    views: {},
    template: require('../../fixtures/template_prize.json')
  },
  victim: {
    documents: {},
    revisions: {},
    views: {},
    template: require('../../fixtures/template_victim.json')
  },
  labor: {
    documents: {},
    revisions: {},
    views: {},
    template: require('../../fixtures/template_labor.json')
  },
  project: {
    documents: {},
    revisions: {},
    views: {},
    template: require('../../fixtures/template_project.json')
  }
};

describe('QueryResults', () => {
  let store;
  let graph = new QueryGraph(models);

  it('renders', async () => {
    store = mockStore({
      magma: {models},
      janus: {projects: require('../../fixtures/project_names.json')}
    });

    let mockGraphState = {
      graph,
      rootModel: 'monster'
    };

    let mockColumnState = {
      columns: [
        {
          model_name: 'monster',
          attribute_name: 'name',
          display_label: 'monster.name',
          slices: []
        },
        {
          model_name: 'prize',
          attribute_name: 'name',
          display_label: 'prize.name',
          slices: [
            {
              modelName: 'prize',
              attributeName: 'name',
              operator: '::equals',
              operand: 'Athens',
              attributeType: 'text'
            }
          ]
        }
      ]
    };

    let mockWhereState = {
      orRecordFilterIndices: [],
      recordFilters: [
        {
          modelName: 'labor',
          attributeName: 'year',
          operator: '::equals',
          operand: 2,
          anyMap: {},
          attributeType: 'number'
        }
      ]
    };

    const {asFragment} = render(<QueryResults />, {
      wrapper: querySpecWrapper({
        mockGraphState,
        mockColumnState,
        mockWhereState,
        mockResultsState: defaultQueryResultsParams,
        store
      })
    });

    await waitFor(() => screen.getByText('Nest matrices'));

    expect(asFragment()).toMatchSnapshot();

    expect(screen.getByText('monster.name')).toBeTruthy();
    expect(screen.getByText('prize.name')).toBeTruthy();
  });

  it('expands and nests matrix columns', async () => {
    store = mockStore({
      magma: {models},
      janus: {projects: require('../../fixtures/project_names.json')}
    });

    let mockGraphState = {
      graph,
      rootModel: 'monster'
    };

    let mockColumnState = {
      columns: [
        {
          model_name: 'monster',
          attribute_name: 'name',
          display_label: 'monster.name',
          slices: []
        },
        {
          model_name: 'prize',
          attribute_name: 'name',
          display_label: 'prize.name',
          slices: [
            {
              modelName: 'prize',
              attributeName: 'name',
              operator: '::equals',
              operand: 'Athens',
              attributeType: 'text'
            }
          ]
        },
        {
          model_name: 'labor',
          attribute_name: 'contributions',
          display_label: 'labor.contributions',
          slices: [
            {
              modelName: 'labor',
              attributeName: 'contributions',
              operator: '::slice',
              operand: 'Athens,Sparta',
              attributeType: 'matrix'
            }
          ]
        }
      ]
    };

    let mockWhereState = {
      orRecordFilterIndices: [],
      recordFilters: [
        {
          modelName: 'labor',
          attributeName: 'year',
          operator: '::equals',
          operand: 2,
          anyMap: {},
          attributeType: 'number'
        }
      ]
    };

    const {asFragment} = render(<QueryResults />, {
      wrapper: querySpecWrapper({
        mockGraphState,
        mockColumnState,
        mockWhereState,
        mockResultsState: defaultQueryResultsParams,
        store
      })
    });

    await waitFor(() => screen.getByText('Nest matrices'));

    expect(asFragment()).toMatchSnapshot();

    expect(screen.getByText('labor.contributions.Athens')).toBeTruthy();
    expect(screen.getByText('labor.contributions.Sparta')).toBeTruthy();

    fireEvent.click(screen.getByText('Nest matrices'));

    expect(screen.getByText('labor.contributions')).toBeTruthy();
    expect(() => screen.getByText('labor.contributions.Athens')).toThrowError();
    expect(() => screen.getByText('labor.contributions.Sparta')).toThrowError();
  });
});
