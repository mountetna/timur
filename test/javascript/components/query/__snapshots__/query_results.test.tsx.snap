// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryResults expands and nests matrix columns 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-resultsPane MuiGrid-container"
  >
    <div
      class="MuiGrid-root makeStyles-result MuiGrid-item"
    >
      <div>
        <div
          class="cm-editor ͼ1 ͼ2 "
        >
          <div
            aria-live="polite"
            style="position: absolute; top: -10000px;"
          />
          <div
            class="cm-scroller"
            tabindex="-1"
          >
            <div
              aria-multiline="true"
              autocapitalize="off"
              autocorrect="off"
              class="cm-content cm-lineWrapping"
              contenteditable="false"
              role="textbox"
              spellcheck="false"
              style="tab-size: 4;"
              translate="no"
            >
              <div
                class="cm-line"
              >
                query: 
              </div>
              <div
                class="cm-line"
              >
                user_columns: [
                <span
                  class="ͼd"
                >
                  "monster.name"
                </span>
                ,
                <span
                  class="ͼd"
                >
                  "monster.name"
                </span>
                ,
                <span
                  class="ͼd"
                >
                  "prize.name"
                </span>
                ,
                <span
                  class="ͼd"
                >
                  "labor.contributions"
                </span>
                ]
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-12"
    >
      <div
        class="MuiGrid-root makeStyles-config MuiGrid-container MuiGrid-item MuiGrid-justify-xs-flex-end"
      >
        <div
          class="MuiTypography-root makeStyles-container MuiTypography-body1"
        >
          <label
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              Nest matrices
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <span
                class="MuiSwitch-root WithStyles(ForwardRef(Switch))-root"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root MuiSwitch-switchBase WithStyles(ForwardRef(Switch))-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked MuiSwitch-checked WithStyles(ForwardRef(Switch))-checked"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      checked=""
                      class="PrivateSwitchBase-input MuiSwitch-input"
                      name="expand-matrices-query"
                      type="checkbox"
                      value=""
                    />
                    <span
                      class="MuiSwitch-thumb WithStyles(ForwardRef(Switch))-thumb"
                    />
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </span>
                <span
                  class="MuiSwitch-track WithStyles(ForwardRef(Switch))-track"
                />
              </span>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              Expand matrices
            </div>
          </label>
        </div>
        <div
          class="MuiTypography-root makeStyles-container MuiTypography-body1"
        >
          <label
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              Nested
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <span
                class="MuiSwitch-root WithStyles(ForwardRef(Switch))-root"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root MuiSwitch-switchBase WithStyles(ForwardRef(Switch))-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked MuiSwitch-checked WithStyles(ForwardRef(Switch))-checked"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      checked=""
                      class="PrivateSwitchBase-input MuiSwitch-input"
                      name="flatten-query"
                      type="checkbox"
                      value=""
                    />
                    <span
                      class="MuiSwitch-thumb WithStyles(ForwardRef(Switch))-thumb"
                    />
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </span>
                <span
                  class="MuiSwitch-track WithStyles(ForwardRef(Switch))-track"
                />
              </span>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              Flattened
            </div>
          </label>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-flex-end MuiGrid-justify-xs-flex-end"
      >
        <div
          class="MuiTablePagination-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
          >
            <div
              class="MuiTablePagination-spacer"
            />
            <p
              class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
              id="mui-50000"
            >
              Rows per page:
            </p>
            <div
              class="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
            >
              <div
                aria-haspopup="listbox"
                aria-labelledby="mui-50000 mui-50000"
                class="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
                id="mui-50000"
                role="button"
                tabindex="0"
              >
                10
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput"
                tabindex="-1"
                value="10"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
            <p
              class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
            >
              0-0 of 0
            </p>
            <div
              class="MuiTablePagination-actions"
            >
              <button
                aria-label="Previous page"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-disabled MuiButtonBase-disabled"
                disabled=""
                tabindex="-1"
                title="Previous page"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                    />
                  </svg>
                </span>
              </button>
              <button
                aria-label="Next page"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-disabled MuiButtonBase-disabled"
                disabled=""
                tabindex="-1"
                title="Next page"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiTableContainer-root"
      >
        <table
          aria-label="result table"
          class="MuiTable-root makeStyles-table"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                scope="col"
              >
                monster.name
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                scope="col"
              >
                prize.name
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                scope="col"
              >
                labor.contributions.Athens
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                scope="col"
              >
                labor.contributions.Sparta
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          />
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`QueryResults renders 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-resultsPane MuiGrid-container"
  >
    <div
      class="MuiGrid-root makeStyles-result MuiGrid-item"
    >
      <div>
        <div
          class="cm-editor ͼ1 ͼ2 "
        >
          <div
            aria-live="polite"
            style="position: absolute; top: -10000px;"
          />
          <div
            class="cm-scroller"
            tabindex="-1"
          >
            <div
              aria-multiline="true"
              autocapitalize="off"
              autocorrect="off"
              class="cm-content cm-lineWrapping"
              contenteditable="false"
              role="textbox"
              spellcheck="false"
              style="tab-size: 4;"
              translate="no"
            >
              <div
                class="cm-line"
              >
                query: 
              </div>
              <div
                class="cm-line"
              >
                user_columns: [
                <span
                  class="ͼd"
                >
                  "monster.name"
                </span>
                ,
                <span
                  class="ͼd"
                >
                  "monster.name"
                </span>
                ,
                <span
                  class="ͼd"
                >
                  "prize.name"
                </span>
                ]
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-12"
    >
      <div
        class="MuiGrid-root makeStyles-config MuiGrid-container MuiGrid-item MuiGrid-justify-xs-flex-end"
      >
        <div
          class="MuiTypography-root makeStyles-container MuiTypography-body1"
        >
          <label
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              Nest matrices
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <span
                class="MuiSwitch-root WithStyles(ForwardRef(Switch))-root"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root MuiSwitch-switchBase WithStyles(ForwardRef(Switch))-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked MuiSwitch-checked WithStyles(ForwardRef(Switch))-checked"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      checked=""
                      class="PrivateSwitchBase-input MuiSwitch-input"
                      name="expand-matrices-query"
                      type="checkbox"
                      value=""
                    />
                    <span
                      class="MuiSwitch-thumb WithStyles(ForwardRef(Switch))-thumb"
                    />
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </span>
                <span
                  class="MuiSwitch-track WithStyles(ForwardRef(Switch))-track"
                />
              </span>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              Expand matrices
            </div>
          </label>
        </div>
        <div
          class="MuiTypography-root makeStyles-container MuiTypography-body1"
        >
          <label
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              Nested
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <span
                class="MuiSwitch-root WithStyles(ForwardRef(Switch))-root"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root MuiSwitch-switchBase WithStyles(ForwardRef(Switch))-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked MuiSwitch-checked WithStyles(ForwardRef(Switch))-checked"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      checked=""
                      class="PrivateSwitchBase-input MuiSwitch-input"
                      name="flatten-query"
                      type="checkbox"
                      value=""
                    />
                    <span
                      class="MuiSwitch-thumb WithStyles(ForwardRef(Switch))-thumb"
                    />
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </span>
                <span
                  class="MuiSwitch-track WithStyles(ForwardRef(Switch))-track"
                />
              </span>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              Flattened
            </div>
          </label>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-flex-end MuiGrid-justify-xs-flex-end"
      >
        <div
          class="MuiTablePagination-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
          >
            <div
              class="MuiTablePagination-spacer"
            />
            <p
              class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
              id="mui-50000"
            >
              Rows per page:
            </p>
            <div
              class="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
            >
              <div
                aria-haspopup="listbox"
                aria-labelledby="mui-50000 mui-50000"
                class="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
                id="mui-50000"
                role="button"
                tabindex="0"
              >
                10
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput"
                tabindex="-1"
                value="10"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
            <p
              class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
            >
              0-0 of 0
            </p>
            <div
              class="MuiTablePagination-actions"
            >
              <button
                aria-label="Previous page"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-disabled MuiButtonBase-disabled"
                disabled=""
                tabindex="-1"
                title="Previous page"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                    />
                  </svg>
                </span>
              </button>
              <button
                aria-label="Next page"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-disabled MuiButtonBase-disabled"
                disabled=""
                tabindex="-1"
                title="Next page"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiTableContainer-root"
      >
        <table
          aria-label="result table"
          class="MuiTable-root makeStyles-table"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                scope="col"
              >
                monster.name
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                scope="col"
              >
                prize.name
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          />
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
